### PR TITLE
Add "Next Scan" timestamp to folder view #3586

### PR DESCRIPTION
**Pull request by hfichtenberger**
### Purpose

Adds an entry _Next Scan_ to each folder in the GUI that shows when the next scan will happen (similar to the _Last Scan_ entry) as requested by #3586.
### Testing

I didn't find any tests I could hook into. Currently, I don't think I can write a test in the way I should to test this feature from scratch (i.e., set up the environment, create a dummy share and check whether the time shows right) because I'm just starting to read the syncthing code. I checked that the next scan is correct on my installation manually. I also checked that if the timestamp is not greater than the last scan's timestamp the entry is hidden.

However, it seems to me that `func (f *folderScanner) Reschedule() time.Time` is called just after the StateChange event in `func (m *Model) internalScanFolderSubdirs(folder string, subDirs []string) error` is fired. I think that, in theory, the GUI could query the new folder state just before the next scan timestamp is updated. I would suggest to fire another StateChanged event right after updating the next scan timestamp in the folderState object. It seems to me that there is no intended way to do this without actually changing the state to another value (see `func (s *stateTracker) setState(newState folderState)`). What would be the right way to do this?
### Screenshots

firefox_2016-09-27_22-29-32
### Documentation

A new screenshot of the GUI may be added to docs/intro/gui.rst (the current one seems to be outdated anyway).

### CODE REVIEW

File: lib/stats/folder.go [hunk 1]
```diff
@@ -15,6 +15,7 @@ import (
 type FolderStatistics struct {
 	LastFile LastFile  `json:"lastFile"`
 	LastScan time.Time `json:"lastScan"`
+	NextScan time.Time `json:"nextScan"`
```
Comment (AudriusButkevicius) [inline lib/stats/folder.go]: This number will be a lie after restarting most likely.

File: gui/default/index.html [hunk 2]
```diff
@@ -393,6 +393,12 @@ <h4 class="panel-title">
                         <span>{{folderStats[folder.id].lastScan | date:'yyyy-MM-dd HH:mm:ss'}}</span>
                       </td>
                     </tr>
+                    <tr ng-if="folderStats[folder.id].nextScanDelta > 0">
+                      <th><span class="fa fa-fw fa-clock-o"></span>&nbsp;<span translate>Next Scan</span></th>
+                      <td class="text-right">
+                        <span>{{folderStats[folder.id].nextScan | date:'yyyy-MM-dd HH:mm:ss'}}</span>
+                      </td>
+                    </tr>
```
Comment (AudriusButkevicius) [inline gui/default/index.html]: Can we just have next time as a tooltip for the last scan? UI space is precious.

**Comment by calmh** [comment]
Haven't looked at the code yet, but to answer the question in the description... we should send the next scan time in the state change to "idle". That may mean we need to calculate it earlier, I don't know. We should also consider that an ongoing pull is going to potentially delay the next scan, just so that this doesn't get confusing in the GUI.

**Comment by uok** [comment]
@hfichtenberger thanks for working on that improvement! Suggestion for icon: fa-angle-double-right or fa-forward

**Comment by hfichtenberger** [comment]
@calmh As I said I'm not too familiar with the high-level philosophy of the code but I guess the next scan time shouldn't be calculated and set in `func (m *Model) internalScanFolderSubdirs(folder string, subDirs []string)` because this is not the semantic scope of this method. Actually, it's also called for on-demand scans so that rescheduling might not be intended in this case. I can imagine too possible solutions: One would be to pull calls to `func (s *stateTracker) setState(newState folderState)` into methods on a higher level, probably into `func (f *rwFolder) Serve()` and `func (f *roFolder) Serve()` (which means that the actual scanning would be detached from the management of the folder state, so one would have to check the call graph very carefully for all calls to the scan methods). The other solution would be to introduced a method like `func (s *stateTracker) updateState()`, which could fire a StateChanged "from x to x" event that indicates a change of the _state metadata_. This one seems less invasive to me with the downside of firing to events.

File: lib/stats/folder.go (hunk 1 above)
Comment (hfichtenberger) [inline lib/stats/folder.go]: I see that this is saved into some LevelDB file. As the next scan time is calculated in any case after restarting, I guess that this shouldn't be saved in the first place. Do you know by chance where this serialization happens?

Comment (AudriusButkevicius) [inline lib/stats/folder.go]: So I suggest you wedge it in here:

https://sourcegraph.com/github.com/hfichtenberger/syncthing@faf3a96db78f3a17449d3abbc0ec2c3dd8dab6a8/-/blob/cmd/syncthing/gui.go#L612

So just add a new method to the model m.NextScan("folder") which is potentially stored on the runner, and then returned in the summary.

**Comment by calmh** [comment]
I hear you. Propose a solution. :)

**Comment by calmh** [comment]
Timeout, please feel free to reopen if this gets closer to completion.

**Pull request closed without merging**